core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,360 @@
+                    estimatePartitionWrittenBytesSinceLastRebalance(writerPartitionId.writerId, rowCount));
+            writerPartitionsWithWrittenBytes.get(writerPartitionId.writerId).addOrUpdate(idWithPriority, idWithPriority.writtenBytes);
+        });
+
+        Set<Integer> rebalancedPartitionIds = new HashSet<>();
+
+        // Find skewed partitions and scale them across multiple writers
+        while (true) {
+            // Find the writer with maximum physical written bytes since last rebalance
+            IdWithWrittenBytes maxWriter = maxWriters.poll();
+
+            if (maxWriter == null) {
+                break;
+            }
+
+            // Find the skewness against writer with max physical written bytes since last rebalance
+            List<IdWithWrittenBytes> minSkewedWriters = findSkewedMinWriters(maxWriter.writtenBytes, minWriters);
+            if (minSkewedWriters.isEmpty()) {
+                break;
+            }
+
+            int minWriterIdWherePartitionIsRebalanced = -1;
+            for (IdWithWrittenBytes minSkewedWriter : minSkewedWriters) {
+                Optional<IdWithWrittenBytes> optionalRebalancedPartition = rebalancePartition(
+                        minSkewedWriter.id,
+                        writerPartitionsWithWrittenBytes.get(maxWriter.id),
+                        rebalancedPartitionIds);
+
+                if (optionalRebalancedPartition.isPresent()) {
+                    IdWithWrittenBytes rebalancedPartition = optionalRebalancedPartition.get();
+                    rebalancedPartitionIds.add(rebalancedPartition.id);
+
+                    // Since a partition is rebalanced from max to min skewed writer, decrease the writtenBytes of max
+                    // writer as well as increase the writtenBytes of min writer.
+                    IdWithWrittenBytes maxWriterWithPriority = new IdWithWrittenBytes(
+                            maxWriter.id,
+                            maxWriter.writtenBytes - (rebalancedPartition.writtenBytes / 2));
+                    IdWithWrittenBytes minWriterWithPriority = new IdWithWrittenBytes(
+                            minSkewedWriter.id,
+                            minSkewedWriter.writtenBytes + (rebalancedPartition.writtenBytes / 2));
+
+                    maxWriters.addOrUpdate(maxWriterWithPriority, maxWriterWithPriority.writtenBytes);
+                    maxWriters.addOrUpdate(minWriterWithPriority, minWriterWithPriority.writtenBytes);
+                    minWriters.addOrUpdate(maxWriterWithPriority, Long.MAX_VALUE - maxWriterWithPriority.writtenBytes);
+                    minWriters.addOrUpdate(minWriterWithPriority, Long.MAX_VALUE - minWriterWithPriority.writtenBytes);
+
+                    minWriterIdWherePartitionIsRebalanced = minSkewedWriter.id;
+                    break;
+                }
+            }
+
+            // Add all the min skewed writers back to the writtenBytes queue except the one where a new partition is rebalanced
```

Comment:
why do we have to make an exception?
Assuming we improved distribution and make progress (because of `rebalancedPartition.writtenBytes / 2`), we shouldn't have exception like this.